torch/csrc/distributed/rpc/python_functions.cpp

```diff
@@ -171,29 +178,46 @@ PyRRef pyRemotePythonUdf(
     std::string& pickledPythonUDF,
     std::vector<torch::Tensor>& tensors) {
   auto& ctx = RRefContext::getInstance();
-  // TODO: support creating RRefs on a local object.
-  TORCH_INTERNAL_ASSERT(
-      ctx.getWorkerId() != dst.id_,
-      "Does not support creating RRef on self yet.");
-  auto userRRef = ctx.createUserRRef<py::object>(dst.id_);
-
-  auto pythonRemoteCall = c10::guts::make_unique<PythonRemoteCall>(
-      SerializedPyObj(std::move(pickledPythonUDF), std::move(tensors)),
-      userRRef->rrefId().toIValue(),
-      userRRef->forkId().toIValue());
-
-  // set forceGradRecording to true as even if the args does not contain any
-  // tensor, the return value might still contain tensors.
-  auto fm = sendMessageWithAutograd(
-      agent,
-      dst,
-      std::move(*pythonRemoteCall).toMessage(),
-      true /*forceGradRecording*/
-  );
-
-  ctx.addPendingUser(userRRef->forkId(), userRRef);
-  fm->addCallback(finishAcceptUserRRef);
-  return PyRRef(userRRef);
+  if (ctx.getWorkerId() != dst.id_) {
+    auto userRRef = ctx.createUserRRef<py::object>(dst.id_);
+    auto pythonRemoteCall = c10::guts::make_unique<PythonRemoteCall>(
+        SerializedPyObj(std::move(pickledPythonUDF), std::move(tensors)),
+        userRRef->rrefId().toIValue(),
+        userRRef->forkId().toIValue());
+
+    // set forceGradRecording to true as even if the args does not contain any
+    // tensor, the return value might still contain tensors.
+    auto fm = sendMessageWithAutograd(
+        agent,
+        dst,
+        std::move(*pythonRemoteCall).toMessage(),
+        true /*forceGradRecording*/
+    );
+
+    ctx.addPendingUser(userRRef->forkId(), userRRef);
+    fm->addCallback(finishAcceptUserRRef);
+    return PyRRef(userRRef);
+  } else {
+    auto ownerRRef = ctx.createOwnerRRef<py::object>(true /* retainInContext*/);
```

Comment:
retainInContext is hard to understand. would it make sense to make this a separate function call?

Reply:
Sure, I am working on refactoring this PR, will send out an update today.